File: packages/android/fernlink-sdk/src/main/kotlin/xyz/fernlink/sdk/ble/ProofStore.kt
package xyz.fernlink.sdk.ble

import java.util.concurrent.ConcurrentLinkedQueue

/**
 * Store-and-forward buffer for outbound verification requests.
 *
 * When no BLE peers are reachable, requests are queued here.
 * GattClientManager drains the queue each time a new peer connects.
 */
internal class ProofStore {

    data class PendingRequest(
        val txSignature: String,
        val statusByte: Byte,
        val slot: Long,
        val blockTime: Long,
    )

    private val queue = ConcurrentLinkedQueue<PendingRequest>()

    val size: Int get() = queue.size
    val isEmpty: Boolean get() = queue.isEmpty()

    fun enqueue(req: PendingRequest) {
        // Cap at 64 entries to bound memory use on long-offline devices
        if (queue.size < 64) queue.add(req)
    }

    /** Drain and return all buffered requests, clearing the queue. */
    fun drain(): List<PendingRequest> {
        val out = mutableListOf<PendingRequest>()
        while (true) out.add(queue.poll() ?: break)
        return out
    }

    fun clear() = queue.clear()
}

File: packages/ios/Sources/FernlinkSDK/ProofStore.swift
import Foundation

/// Store-and-forward buffer for outbound verification requests.
/// Queues requests when no BLE peers are reachable; drained on reconnect.
final class ProofStore {
    struct PendingRequest {
        let txSignature: String
        let commitment:  String
        let ttl:         Int
    }

    private let lock  = NSLock()
    private var queue = [PendingRequest]()
    private let cap   = 64

    var isEmpty: Bool { lock.withLock { queue.isEmpty } }
    var count:   Int  { lock.withLock { queue.count } }

    func enqueue(_ req: PendingRequest) {
        lock.withLock { if queue.count < cap { queue.append(req) } }
    }

    func drain() -> [PendingRequest] {
        lock.withLock {
            let all = queue
            queue = []
            return all
        }
    }

    func clear() { lock.withLock { queue = [] } }
}
